From fedd1750923dd2019a1ffbbbba1f5abf1e380aa3 Mon Sep 17 00:00:00 2001
From: Changqing Li <changqing.li@windriver.com>
Date: Mon, 7 Apr 2025 07:35:24 +0000
Subject: [PATCH] strutl.cc: the filename can't be longer than 255

The URItoFileName translates the path into the filename, but the
filename can't be longer than 255 according to
/usr/include/linux/limits.h.

Truncate it when it is longer than 240 (leave some spaces for
".Packages" and "._Release" suffix)

Upstream-Status: Submitted [https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1073591]

Signed-off-by: Robert Yang <liezhi.yang@windriver.com>
Signed-off-by: Changqing Li <changqing.li@windriver.com>
---
 apt-pkg/contrib/strutl.cc | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/apt-pkg/contrib/strutl.cc b/apt-pkg/contrib/strutl.cc
index c0a1cbc..695037b 100644
--- a/apt-pkg/contrib/strutl.cc
+++ b/apt-pkg/contrib/strutl.cc
@@ -566,7 +566,12 @@ string URItoFileName(const string &URI)
    // "\x00-\x20{}|\\\\^\\[\\]<>\"\x7F-\xFF";
    string NewURI = QuoteString(U,"\\|{}[]<>\"^~_=!@#$%^&*");
    replace(NewURI.begin(),NewURI.end(),'/','_');
-   return NewURI;
+
+   // Truncate from the head when it is longer than 240
+   if(NewURI.length() > 240)
+       return NewURI.substr(NewURI.length() - 240, NewURI.length() - 1);
+   else
+       return NewURI;
 }
 									/*}}}*/
 // Base64Encode - Base64 Encoding routine for short strings		/*{{{*/
-- 
2.48.1

